File: luiz/ba1h.py
#! /usr/bin/env python


from ba1g import hamming


def approximate_pattern_match(text, pattern, distance):
    matches = []
    for i in range(len(text) - len(pattern)):
        if hamming(text[i : i + len(pattern)], pattern) <= distance:
            matches.append(i)
    return matches


assert approximate_pattern_match(
    "CGCCCGAATCCAGAACGCATTCCCATATTTCGGGACCACTGGCCTCCACGGTACGGACGTCAATCAAATGCCTAGCGGCTTGTGGTTTCTCCTACGCTCC",
    "ATTCTGGA",
    3,
) == [6, 7, 26, 27, 78]


if __name__ == "__main__":
    with open("data/rosalind_ba1h.txt", "r") as dataset:
        pattern = dataset.readline().rstrip()
        text = dataset.readline().rstrip()
        dist = int(dataset.readline().rstrip())
    print(*approximate_pattern_match(text, pattern, dist), sep=" ")

File: luiz/src/ba1h.rs
use std::env;
use std::error::Error;
use std::fs;
use std::str;

use bioinformatics_algorithms::hamming;

pub fn approximate_pattern_matches(text: &[u8], pattern: &str, distance: usize) -> Vec<usize> {
    text.windows(pattern.len())
        .enumerate()
        .filter_map(|(i, x)| {
            if hamming(str::from_utf8(x).unwrap(), pattern) <= distance {
                Some(i)
            } else {
                None
            }
        })
        .collect()
}

fn main() -> Result<(), Box<dyn Error>> {
    assert_eq!(
        approximate_pattern_matches(b"CGCCCGAATCCAGAACGCATTCCCATATTTCGGGACCACTGGCCTCCACGGTACGGACGTCAATCAAATGCCTAGCGGCTTGTGGTTTCTCCTACGCTCC", "ATTCTGGA", 3),
        vec![6, 7, 26, 27, 78]
    );

    let input: String = env::args().nth(1).expect("Input data file missing");
    let data = fs::read_to_string(input)?;
    let mut lines = data.lines();

    let pattern = lines.next().unwrap();
    let text = lines.next().unwrap();
    let d = lines.next().unwrap().parse().unwrap();

    for word in approximate_pattern_matches(text.as_bytes(), pattern, d) {
        print!("{} ", word);
    }
    println!();

    Ok(())
}
